### ISSUE TITLE
Unnecessary Paging of Results in psql

### ISSUE THREAD

**Issue by aksuska** [PostgresApp/PostgresApp#607]
Select results always cause pagination pauses regardless of result count and whether pagination is even called for. Version 2.4.1 on Mac OS 10.15.7.

**Comment by jakob** [comment]
I don't see any pagination on my machine. Do you maybe have a `.psqlrc` file with custom settings in your home directory, or did you set any psql-related environment variables?

**Comment by aksuska** [comment]
No .psqlrc anywhere on my machine; only pguser & pgpassword are set in ENV. Paging is on by default, IIRC, so you must have paging turned off. It is possible that the difference in behavior is due to an OS change in the "less" command, or a change in default options from inside psql. If I set PAGER to "less -XE" then it behaves as I am used to. Psql may have to at least call "less -X" to get default paging behavior when paging is on, at least on recent versions of Mac OS X.

**Comment by jakob** [comment]
Sorry, I meant I don't see unnecessary paging.

The following doesn't cause pagination on my machine:

```
jakob=# select * from generate_series(1,5);
```

but this command does:

```
jakob=# select * from generate_series(1,500);
```


It looks like psql is just calling `more`. (I checked by calling `ps` while psql is paginating)

This machine is on macOS 10.15.7 with Postgres.app 2.4.1 as well. I'm using Terminal and the default zsh, but I see the same behavior if I start a new shell with `env -i sh`, so it's unlikely that my environment is the problem. I don't have a `.psqlrc` file.

`which psql` returns `/Applications/Postgres.app/Contents/Versions/latest/bin/psql`, so I am using Postgres.app's psql, and `which more` points to the default install of more `/usr/bin/more` that comes with macOS. `psql -V` returns `psql (PostgreSQL) 13.1`.

Are you seeing the same behavior, or something different?

**Comment by aksuska** [comment]
I am also on 10.15.7, and more -V shows "less 487 (POSIX regular expressions)", which is the same as less -V. I am using /Applications/Postgres.app/Contents/Versions/9.6/bin/psql because I need older compatibility. Digging a little deeper, it appears that psql is acting as if \pset pager is set to "always" rather than "on". Setting it to "on" doesn't change behavior, i.e., it always invokes the pager even for a few lines of result. Perhaps this was a bug with that version of psql, but in any case it is  not behaving as documented. Another workaround is to set "pager_min_lines" but is less useful when the terminal window height is changed.

**Comment by jakob** [comment]
I just tried again, with `/Applications/Postgres.app/Contents/Versions/9.6/bin/psql`, on a different machine with macOS 10.15.7, and the command `select * from generate_series(1,5);` still doesn't call the pager.

The only way I can get psql to use the pager for this query is to make the terminal window height smaller than 8 lines.

Maybe psql for some reason thinks you have a tiny terminal? I've been testing in Terminal.app with zsh. What Terminal / shell are you using?

**Comment by aksuska** [comment]
Terminal.app with bash. Not sure how psql is getting terminal sizes, but bash knows:
$ tput lines
57

**Comment by jakob** [comment]
OK, I think I'm at the end of my wits. I have no idea why it doesn't work, and I can't reproduce the issue on my machines. At this point I unfortunately don't have any time to follow up on this any further.

Maybe you could try searching PostgreSQL mailing lists if someone encountered a similar issue, or you could try different psql installations to determine if it's something that only happens with Postgres.app.

If you figure it out, please let us know!

**Comment by aksuska** [comment]
I appreciate your effort to resolve the issue. When I have some time I might test with some compilation from source and see if I can find any difference. In the meantime I have a workaround. Thanks!

**Comment by tbussmann** [comment]
I have not been able to reproduce this so far either, but while working on a different machine (macOS 11.2, M1, experimental builds) at least I can now confirm there is something fishy with result paging. I have not yet investigated it in detail, but I can reproduce an unexpected behaviour. Reopening.

**Comment by tbussmann** [comment]
Further testing showed, that the seemingly unnecessary paging happens if the output is to *wide*, not to *tall* for the current window. For example a `SELECT 1;` will not show the pager but a `SELECT version();` will.  It seems, this behaviour it normal even in other OS, per man page: "when the output [..] will not fit on the screen." The default pager of psql on macOS is `more` unless you set the environment variables `PSQL_PAGER` or `PAGER`. Setting this to `more -S`, `more -E` or even `less -XSFe` looks sensible to me. But this is up to the user.

If you set `\pset format wrapped` instead if it's default `aligned` the pagination will be limited to tall output and individual cells will be wrapped where necessary, `\pset expanded auto` aka `\x auto` is another approach that can be useful.

In summary, I conclude this behaviour is expected and not a bug.
